SPEC.md

```diff
@@ -0,0 +1,334 @@
+      callbacks.
+```
+
+### OperationInfo
+
+The `OperationInfo` object, retrieved from the get operation info API, MUST adhere to the given schema:
+
+```yaml
+type: object
+properties:
+  state:
+    enum:
+      - succeeded
+      - failed
+      - canceled
+      - running
+    description: |
+      Describes the current state of an operation.
+```
+
+## Endpoint Descriptions
+
+### Start Operation
+
+Start an arbirary length operation.
+The response of the operation may be delivered synchronously (inline), or asynchronously, via a provided callback or the
+[Get Operation Result](#get-operation-result) endpoint.
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}`
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}`
+
+#### Method
+
+`POST`
+
+#### Query Parameters
+
+- `callback_url`: Optional. If the operation is asynchronous and the handler supports callbacks, it should invoke this URL once the operation's result is available.
+
+#### Request Body
+
+The body may contain arbitrary data. Headers should specify content type and encoding.
+
+#### Response Codes
+
+- `200 OK`: Operation completed successfully.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded`
+
+  **Body**: Arbitrary data conveying the operation's result. Headers should specify content type and encoding.
+
+- `204 No Content`: Operation completed successfully with an empty result.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded`
+
+  **Body**: Empty.
+
+- `482 Operation Failed`: Operation completed as `failed` or `canceled`.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**: A JSON serialized [`Failure`](#failure) object.
+
+- `201 Created`: Operation was started an will completed asynchronously.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+
+  **Body**: A JSON serialized [`OperationStarted`](#operationstarted) object.
+
+### Cancel Operation
+
+Request to cancel an operation.
+The operation may later complete as canceled or any other outcome.
+Handlers should ignore multiple cancelations of the same operation and return successfully if cancelation was already
+requested.
+
+#### Method
+
+`POST`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}/cancel`
+
+#### Response Codes
+
+- `202 Accepted`: Cancelation request accepted.
+
+  **Body**: Unspecified.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+  **Body**: Unspecified.
+
+### Get Operation Result
+
+Retrieve operation result.
+
+#### Method
+
+`GET`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}/result`
+
+#### Query Parameters
+
+- `wait_deadline`: Optional. ISO 8601 date indicating the waiting period for a result.
+  If not specified, and the operation is still running, the request should resolve immediately with the current
+  operation status.
```

Comment:
I think it should fail with a not-done-yet. The only success result to this call should ever be a success result with the result.

Reply:
I get the part about success being reserved for completion, not sure about using failure for `running` state.

Reply:
But it is a failure of getting a response, not a failure of the operation. Like a 404. Can use different status code if you want (though not 408, my reading is showing me that has certain HTTP connection lifetime expectations).